repo/format/format_blob_cache.go

```diff
@@ -12,133 +14,161 @@ import (
-		return nil, time.Time{}, errors.Wrapf(err, "error getting %s blob", blobID)
-	}
+	//nolint:gosec
+	data, err := os.ReadFile(cachedFile)
+
+	return data, cacheMTime, err == nil
+}
 
-	if cacheEnabled {
-		if err := atomicfile.Write(cachedFile, b.Bytes().Reader()); err != nil {
-			log(ctx).Warnf("unable to write cache: %v", err)
+func (c *onDiskCache) Put(ctx context.Context, blobID blob.ID, data []byte) (time.Time, error) {
+	cachedFile := filepath.Join(c.cacheDirectory, string(blobID))
+
+	// optimistically assume cache directory exist, create it if not
+	if err := atomicfile.Write(cachedFile, bytes.NewReader(data)); err != nil {
+		if err := os.MkdirAll(c.cacheDirectory, cache.DirMode); err != nil && !os.IsExist(err) {
+			return time.Time{}, errors.Wrap(err, "unable to create cache directory")
 		}
-	}
 
-	return b.ToByteSlice(), clock.Now(), nil
-}
+		if err := cachedir.WriteCacheMarker(c.cacheDirectory); err != nil {
+			return time.Time{}, errors.Wrap(err, "unable to write cache directory marker")
+		}
 
-// ReadAndCacheDecodedRepositoryConfig reads `kopia.repository` blob, potentially from cache and decodes it.
-func ReadAndCacheDecodedRepositoryConfig(ctx context.Context, st blob.Storage, password, cacheDir string, validDuration time.Duration) (ufb *DecodedRepositoryConfig, err error) {
-	ufb = &DecodedRepositoryConfig{}
+		if err := atomicfile.Write(cachedFile, bytes.NewReader(data)); err != nil {
+			return time.Time{}, errors.Wrapf(err, "unable to write to cache: %v", string(blobID))
+		}
+	}
 
-	ufb.KopiaRepositoryBytes, ufb.CacheMTime, err = ReadAndCacheRepositoryBlobBytes(ctx, st, cacheDir, KopiaRepositoryBlobID, validDuration)
+	cst, err := os.Stat(cachedFile)
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to read format blob")
+		return time.Time{}, errors.Wrap(err, "unable to open cache file")
 	}
 
-	if err = cachedir.WriteCacheMarker(cacheDir); err != nil {
-		return nil, errors.Wrap(err, "unable to write cache directory marker")
-	}
+	return cst.ModTime(), nil
+}
 
-	ufb.KopiaRepository, err = ParseKopiaRepositoryJSON(ufb.KopiaRepositoryBytes)
-	if err != nil {
-		return nil, errors.Wrap(err, "can't parse format blob")
+func (c *onDiskCache) Remove(ctx context.Context, ids []blob.ID) {
+	for _, blobID := range ids {
+		fname := filepath.Join(c.cacheDirectory, string(blobID))
+		log(ctx).Infof("deleting %v", fname)
+
+		if err := os.Remove(fname); err != nil && !os.IsNotExist(err) {
+			log(ctx).Debugf("unable to remove cached repository format blob: %v", err)
+		}
 	}
+}
 
-	ufb.KopiaRepositoryBytes, err = addFormatBlobChecksumAndLength(ufb.KopiaRepositoryBytes)
-	if err != nil {
-		return nil, errors.Errorf("unable to add checksum")
+// NewDiskCache returns on-disk blob cache.
+func NewDiskCache(cacheDir string) BlobCache {
+	return &onDiskCache{cacheDir}
+}
+
+// NewMemoryBlobCache returns in-memory blob cache.
+func NewMemoryBlobCache(timeNow func() time.Time) BlobCache {
```

Comment:
Is it necessary to include the BlobCache in this PR?  If so, why?

Reply:
It's a micro-abstraction just for caching the two blobs  - we have 3 strategies, in-memory, null and on-disk. That's part of the refactoring, the previous approach assumed disk and was generally fragile.